xarray/plot/utils.py

```diff
@@ -98,6 +98,30 @@ def _build_discrete_cmap(cmap, levels, extend, filled):
     # copy the old cmap name, for easier testing
     new_cmap.name = getattr(cmap, "name", cmap)
 
+    # copy colors to use for bad, under, and over values in case they have been
+    # set to non-default values
+    try:
+        # matplotlib<3.2 only uses bad color for masked values
+        bad = cmap(np.ma.masked_invalid([np.nan]))[0]
```

Comment:
Can't you do

```suggestion
        bad = cmap(np.ma.masked_invalid(np.nan))
```
?

Reply:
It seems like I can. I'd just used the same form as mwaskom/seaborn#1948, but don't know why that form was used originally...

Reply:
Maybe because they use `npt.assert_array_equal` to test for equality - `cmap(np.ma.masked_invalid([np.nan]))` returns an numpy array and `cmap(np.ma.masked_invalid(np.nan))` a tuple.

Reply:
Checking this, I get an error from this
```
>>> cmap = copy(mpl.cm.viridis)
>>> cmap(numpy.ma.masked_invalid(numpy.nan))                               
/home/john/.local/lib/python3.6/site-packages/matplotlib/colors.py:509: UserWarning: Warning: converting a masked element to nan.
  xa = np.array([X])
```
and I get different results from the original and the changed version:
```
In [11]: cmap(numpy.ma.masked_invalid(numpy.nan))                               
/home/john/.local/lib/python3.6/site-packages/matplotlib/colors.py:509: UserWarning: Warning: converting a masked element to nan.
  xa = np.array([X])
Out[11]: (0.267004, 0.004874, 0.329415, 1.0)
```
```
In [13]: cmap(numpy.ma.masked_invalid([numpy.nan]))[0]                          
Out[13]: array([0., 0., 0., 0.])
```
And the second one here (the original form from mwaskom/seaborn#1948) is correct
```
In [10]: cmap._rgba_bad                                                         
Out[10]: (0.0, 0.0, 0.0, 0.0)
```
so I will get rid of the commit changing this.

Reply:
Ok, weird - sorry for sending you down the wrong path!

Reply:
Slightly worrying: the incorrect change didn't cause a test fail. I'll add another test case that would have caught this.